### PR TITLE
Test/update options

### PR DESCRIPTION
**Pull request by ssp6**
This is the canary tests for update options using:
- useTokenBalance
- useContractReader

As dictated in EPIC #118 

To get those created also had to create tests for useContractReader (#122 ) and more work on useTokenBalance (#76 )

### CODE REVIEW

File: src/functions/hookHelpers.ts [hunk 1]
```diff
@@ -93,3 +97,13 @@ export const processQueryOptions = <GResult>(
   }
   return queryOptions;
 };
+
+export const shouldFailWithMessage = async (failingFunction: () => Promise<any>, errorMessage: string) => {
```
Comment (ShravanSunder) [inline src/functions/hookHelpers.ts]: lets put this in `src/helpers/test-utils/functions` as we don't want to put test helpers in the same area as runtime functions.  we wouldn't want to import chai

File: src/hooks/erc/useTokenBalance.ts [hunk 2]
```diff
@@ -18,10 +18,9 @@ type ERC20 = {
 /**
  * #### Summary
  * Get the balance of an ERC20 token in an address
- * - uses the ethers.Contract object's provider to access the network
  *
  * ##### ✏️ Notes
- * - uses useOnRepetition
+ * - uses the ethers.Contract object's provider to access the network
```
Comment (ShravanSunder) [inline src/hooks/erc/useTokenBalance.ts:23]: ty 🥳

**Review by ShravanSunder — COMMENTED** [review]
🔥 great work!  and happy about comments 🥳
just left one feedback about helper.

this test seems to be failing as well
````
useContractReader Given that a YourContract is deployed Given the setPurpose is called and set with a new value When given opti… given options of block number interval to update; then the hook does not update until that amount of blocks has passed) failed`

testresult.xml
````
AssertionError: expected 'your purpose' to equal 'no purpose'
    at Context.<anonymous> (/home/runner/work/eth-hooks/eth-hooks/test/useContractReader.test.ts:183:48)
```

File: src/functions/hookHelpers.ts (hunk 1 above)
Comment (ssp6) [reply src/functions/hookHelpers.ts]: Oops, I thought this was in test utils, didn't look at the file path correctly

File: test/useContractReader.test.ts [hunk 3]
```diff
@@ -137,65 +140,65 @@ describe('useContractReader', function () {
...
         const finalPurpose = 'purpose final';
         await yourContract?.setPurpose(finalPurpose);
-        await harness.waitForValueToChange(() => harness.result.current[0], defaultBlockWaitOptions);
+        await wrapper.waitForValueToChange(() => wrapper.result.current[0], defaultBlockWaitOptions);
 
-        expect(harness.result.current[0]).to.eql(finalPurpose);
+        expect(wrapper.result.current[0]).to.eql(finalPurpose);
       });
 
       it.only('When given options of block number interval to update; then the hook does not update until that amount of blocks has passed', async () => {
```
Comment (ssp6) [inline test/useContractReader.test.ts]: Going to want to remove `.only` before pushing

File: test/useContractReader.test.ts [hunk 4]
```diff
@@ -103,3 +108,186 @@ describe('useContractReaderUntyped', function () {
...
+    let yourContract: YourContract | undefined;
+    let contractSigner: Signer;
+    let sandbox: sinon.SinonSandbox;
+
+    before(async () => {
+      // setup a contract
+      const wrapper = await wrapperTestSetupHelper();
+      contractSigner = await getHardhatSigner(wrapper.mockProvider, 1);
+      [yourContract] = await setupMockYourContract(contractSigner);
+      // TODO(sean): Remove once solved 'When given options of block number interval' test
```
Comment (ssp6) [inline test/useContractReader.test.ts]: These will need removed as well before pushing, but they're useful to test the result on different start blocks and if they're all uncommented the test still fails 😅

File: test/useContractReader.test.ts [hunk 5]
```diff
@@ -103,3 +108,186 @@ describe('useContractReaderUntyped', function () {
...
+        );
+
+        console.log(await yourContract?.purpose());
+
+        // -- mine blocks up to block when update should occur
+        const [success, updateBlockNumber] = await mineBlockUntil(
+          wrapper.mockProvider,
+          blockIntervalToUpdate,
+          async (currentBlockNumber): Promise<boolean> => {
+            wrapper.mockProvider.blockNumber;
```
Comment (ssp6) [inline test/useContractReader.test.ts]: Is this required for anything?

File: test/useContractReader.test.ts [hunk 6]
```diff
@@ -103,3 +108,186 @@ describe('useContractReaderUntyped', function () {
...
+            if (currentBlockNumber === blockIntervalToUpdate + testStartBockNumber) {
+              await wrapper.waitForNextUpdate(defaultBlockWaitOptions);
+            }
+            return wrapper.result.current[0] === finalPurpose;
+          }
+        );
+
+        expect(success).to.be.true;
+        expect(updateBlockNumber).to.be.equal(testStartBockNumber + 1 + blockIntervalToUpdate);
```
Comment (ssp6) [inline test/useContractReader.test.ts]: I don't think this is true, this is where it fails for me. If the testStartBockNumber is 6 then it's going to update on block 10/11 (updates every 5 blocks) not 12

File: test/useContractReader.test.ts (hunk 5 above)
Comment (ShravanSunder) [reply test/useContractReader.test.ts]: i've changed it

**Review by ShravanSunder — APPROVED** [review]
🔥

**Pull request merged**